crates/filtering/src/madgwick.rs

```diff
@@ -0,0 +1,94 @@
+use nalgebra::{
+    Matrix4, Quaternion, RealField, Scalar, SimdValue, UnitQuaternion, Vector3, Vector4,
+};
+use num_traits::{One, Zero};
+
+#[derive(Debug)]
+pub struct AccelerometerNormZero;
+
+/// Madgwick's filter for orientation estimation.
+///
+/// This implementation is based on the following:
+/// https://github.com/jmagnuson/ahrs-rs
+pub trait Madgwick<T> {
+    fn update_with_imu(
+        &mut self,
+        gyroscope: &Vector3<T>,
+        accelerometer: &Vector3<T>,
+        filter_gain: T,
+        sample_period: T,
+    ) -> Result<(), AccelerometerNormZero>;
+
+    fn update_with_gyro(&mut self, gyroscope: &Vector3<T>, sample_period: T);
+}
+
+impl<T> Madgwick<T> for UnitQuaternion<T>
+where
+    T: Scalar + SimdValue + RealField + One + Zero + Copy,
+{
+    #[allow(non_snake_case)]
+    fn update_with_imu(
+        &mut self,
+        gyroscope: &Vector3<T>,
+        accelerometer: &Vector3<T>,
+        filter_gain: T,
+        sample_period: T,
+    ) -> Result<(), AccelerometerNormZero> {
+        let q = self.as_ref();
```

Comment:
I know this whole function is probably just copied, but please think about proper names for all the variables used here.

Reply:
I actually did not have a look at the algorithm at all... That would be a lot of effort to first understand how it works, to then subsequently name all variables appropriately